### DISCUSSION TITLE
Tutorial for FFD generation

### DISCUSSION THREAD

**Discussion by Mohsin-Bashir** [mdolab/dafoam#652] [General]
Can we please have a video tutorial for generating FFD box for complex geometries, for example, for rotor37 using ICEM CFD? 
I tried a lot to search for such tutorial but I guess it does not exist.

**Comment by friedenhe** [comment]
## Creating FFD Points with ICEM Structured Mesh
	
- Create the geometry for the overall shape of the ffd boundary
	- Use tools in the geometry tab of the toolbar such as points, curves, and volumes
	- This geometry should be slightly larger than your design surface so that ffd points enclose the surface
- Divide the sections of your geometry into parts
	- Right click parts in the design tree and select 'create part', then select specific lines, surfaces or points of the geometry you would like to include in that part
	- Every item placed in the geometry should be included in one of your parts
- Create blocking
	- In the blocking tab of the toolbar, select create block, the recommended type in the options menu is 3D bounding box
	- If you have multiple sections of your geometry with non-uniform shapes, you may want to 'split block' one or more times to add more edges to associate. Or, one can create a 3D Multizone block to accomodate non-uniform geometry
- Associate Edges
	- Again in the blocking tab, select 'associate'
	- For every boundary edge of the geometry, one by one select a boundary edge of your geometry and then a block edge to associate it to an edge of the block
	- Finally, click 'snap project vertices'
- **Optional:** Create 'O-Grid Block' to define internal mesh parameters
	- Under the 'split block' option, select 'o-grid block'
	- Faces, edges, or vertices of the block can be selected for creating the o-grid
- Define Pre-Mesh Parameters
	- Under the blocking tab, select 'pre-mesh params' then the 'edge params' option
	- You will want to define the number of nodes the mesh will be split into for each edge with the 'nodes' option, if you want to keep certain divisions symmetryic for the entire mesh you can select the copy to all parallel edges option. The nodes will become your ffd points.
	- The spacing of nodes can also be selected with the 'mesh law' option
- Construct the structured mesh
	- Under blocking in the design tree, right click pre-mesh and 'convert to multiblock mesh'
	- Select volume as the domain option
- Output Mesh
	- Under the output mesh tab in the toolbar, select 'select solver' and choose Plot3D as the output solver
	- Finally select 'write input' in the output mesh tab of the toolbar, select all or specific domains of the mesh to output. Open the multiblock you saved previously when prompted, and finally select the output type as 'formatted' and *do not* include 'IBlank array' when saving.

**Comment by friedenhe** [comment]
Can you try the above and see if it works?

**Reply by Mohsin-Bashir** [reply]
Finally it worked!! Thank you so much. I have created a YouTube tutorial video about it for my own reference and for others who may find it helpful. https://youtu.be/aFq4mwVrjbc?si=7jNQ-jVEnsN5Djq1

**Comment by friedenhe** [comment]
Thanks very much for sharing the video! I will link this to the FAQ.
